python/semantic_kernel/connectors/ai/onnx/services/onnx_gen_ai_chat_completion.py

```diff
@@ -0,0 +1,184 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+import logging
+import sys
+from collections.abc import AsyncGenerator
+from typing import Any, ClassVar
+
+if sys.version_info >= (3, 12):
+    from typing import override  # pragma: no cover
+else:
+    from typing_extensions import override  # pragma: no cover
+
+
+from semantic_kernel.connectors.ai.chat_completion_client_base import ChatCompletionClientBase
+from semantic_kernel.connectors.ai.onnx.onnx_gen_ai_prompt_execution_settings import OnnxGenAIPromptExecutionSettings
+from semantic_kernel.connectors.ai.onnx.onnx_gen_ai_settings import OnnxGenAISettings
+from semantic_kernel.connectors.ai.onnx.services.onnx_gen_ai_completion_base import OnnxGenAICompletionBase
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.contents import (
+    ChatHistory,
+    ChatMessageContent,
+    ImageContent,
+    StreamingChatMessageContent,
+    TextContent,
+)
+from semantic_kernel.contents.utils.author_role import AuthorRole
+from semantic_kernel.exceptions import ServiceInvalidExecutionSettingsError
+from semantic_kernel.functions import KernelArguments
+from semantic_kernel.functions.kernel_function import TEMPLATE_FORMAT_MAP
+from semantic_kernel.prompt_template import PromptTemplateConfig
+from semantic_kernel.utils.experimental_decorator import experimental_class
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+@experimental_class
+class OnnxGenAIChatCompletion(ChatCompletionClientBase, OnnxGenAICompletionBase):
+    """OnnxGenAI text completion service."""
+
+    prompt_template: PromptTemplateConfig
```

Comment:
I am not sure if we need a prompt template. I can see that the tokenizer only takes in strings. But can we internally parse the chat history to the format that the tokenizer expects?

Using a prompt template that is passed in could be error-prone at the connector level, especially when you are expecting the template to have certain arguments (in your case the `messages` argument).

Reply:
The problem here is that depending on the Model we need to change the format that the tokenizer expects. So for example Phi3 needs something like this `<|system|>Text<|end|><|user|>Text<|end|>`, while Gemma needs something like this `<bos><start_of_turn>Text<end_of_turn>
<start_of_turn>model`.

[Link to Gemma](https://huggingface.co/google/gemma-7b-it) 
[Link to Phi3](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct)

The .Net Implementation currently only works with Phi3 because the Template is Hardcoded.

The idea of the `prompt_template` here is to enable the Onnx Connector for all models and not only for a specific one, to do this i think we need a prompt_template here.

The unoffical default in ChatTemplates/Prompt is to use the messages object in the `open_ai` json format, thats why i added a injection of the messages per default  in `_apply_chat_template`.

Reply:
yeah I really don't like the idea of having prompt templates here, how many variations are there for the format (there might even be some named conventions for particular formats), because we already account for custom serializing the chathistory for a particular connector, so ideally this just uses that mechanism instead, or it has built-in prompts for specific models, and the dev needs to tell it which one to use and match it to the model used. Another factor is that there are 3 different template languages in python (and an additional one in dotnet), so do we support all of those with this or just one or two? etc. whole can of worms is opened when using templates in this way...

Reply:
Ok , understood ! i will create a helper class, implementing common templates, and then the user can select one of these.